src/main/kotlin/org/rust/ide/utils/RustModTabTitleProvider.kt

```diff
@@ -0,0 +1,18 @@
+package org.rust.ide.utils
+
+import com.intellij.openapi.fileEditor.impl.EditorTabTitleProvider
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.vfs.VirtualFile
+import org.rust.lang.core.psi.impl.isRustFile
+
+class RustModTabTitleProvider : EditorTabTitleProvider {
+    val MOD_FILE_NAME = "mod.rs"
+
+    override fun getEditorTabTitle(project: Project, file: VirtualFile): String? {
+        if (file.isRustFile && MOD_FILE_NAME == file.name) {
+            return "${file.parent.name}/${file.name}"
```

Comment:
Aren't you trying to solve a problem that is already solved by IDEA itself? There's a setting: Editor -> General -> Editor Tabs -> Show directory in editor tabs for non-unique file names. If the flag is set, IDEA adds some intelligence into tabs naming. Given the structure provided by @mkaput, it uses `bar/.../mod.rs` and `foo/.../mod.rs` as the tab titles if both files are open.

Or is it something else you're talking about?